File: erlang/lectures/complexity/tree.erl
-module(tree).

-export([empty/0, lookup/2, modify/3, insert/3, delete/2]).

empty() ->
    nil.

lookup(_, nil) ->
    false;
lookup(K, {node, K, V, _, _}) -> 
    V;
lookup(K, {node, L, _, Left, Right}) -> 
    if 
       K < L ->  
           lookup(K, Left);
       true  -> 
           lookup(K, Right)
    end.


modify(_, _, nil) ->
    nil;
modify(K, V, {node, K, _, Left, Right}) -> 
    {node, K, V, Left, Right};
modify(K, V, {node, L, U, Left, Right}) -> 
    if 
       K < L ->  
           {node, L, U, modify(K, V, Left), Right};
       true  -> 
           {node, L, U,  Left, modify(K, V, Right)}
    end.

insert(K, V, nil) ->
    {node, K, V, nil, nil};
insert(K, V, {node, L, U, Left, Right}) -> 
    if 
       K < L ->  
           {node, L, U, insert(K, V, Left), Right};

       true  -> 
           {node, L, U,  Left, insert(K, V, Right)}
    end.

delete(_, nil) -> nil;
delete(K, {node, K, _, Node, nil}) ->  Node;
delete(K, {node, K, _, nil, Node}) ->  Node;
delete(K, {node, K, _, Left, Right}) -> 
      {U, L, Removed} = lift(Left),
      {node, U, L, Removed, Right};
delete(K, {node, U, L, Left, Right}) -> 
      if 
        K < U -> {node, U, L, delete(K, Left), Right};
        true -> {node, U, L, Left, delete(K, Right)}
    end.

lift({node, K, V, Left, nil}) ->  {K, V, Left};
lift({node, K, V, Left, Right}) ->  
    {U, L, Removed} = lift(Right),
    {U, L, {node, K, V, Left, Removed}}.
    

File: lectures/complexity/src/tree.ex
defmodule Tree do

  def empty() do :nil end

  def lookup(:nil, _) do false end

  def lookup({:node, k, v, _, _}, k) do v end

  def lookup({:node, l, _, left, right}, k) do 
    if k < l do
      lookup(left, k)
    else
      lookup(right, k)
    end
  end

  def modify(:nil, _, _) do :nil end
  def modify({:node, k, _, left, right}, k, v) do 
    {:node, k, v, left, right}
  end
  def modify({:node, l, u, left, right}, k, v) do 
    if k < l do
      {:node, l, u, modify(left, k, v), right}
    else
      {:node, l, u,  left, modify(right, k, v)}
    end
  end
    
  def insert(:nil, k, v) do
    {:node, k, v, :nil, :nil}
  end
  def insert({:node, k, _, left, right}, k, v) do
    {:node, k, v, left, right}
  end
  def insert({:node, l, u, left, right}, k, v) do 
    if k < l do
      {:node, l, u, insert(left, k, v), right}
    else
      {:node, l, u,  left, insert(right, k, v)}
    end
  end

  def delete( :nil, _) do :nil end
  def delete({:node, k, _, node, :nil}, k) do  node end
  def delete({:node, k, _, :nil, node}, k) do  node end
  def delete({:node, k, _, left, right}, k) do 
    {u, l, removed} = lift(left)
    {:node, u, l, removed, right}
  end
  def delete({:node, u, l, left, right}, k) do 
    if k < u do
      {:node, u, l, delete(left, k), right}
    else
      {:node, u, l, left, delete(right, k)}
    end
  end
    
  def lift({:node, k, v, left, :nil}) do {k, v, left} end
  def lift({:node, k, v, left, right}) do  
    {u, l, removed} = lift(right)
    {u, l, {:node, k, v, left, removed}}
  end

end 
